watcher.go

```diff
@@ -0,0 +1,130 @@
+	// bailing out of the watch.
+	MaxAttempts int
+}
+
+var (
+	DefaultWatcher = &Watcher{}
+
+	// WatchTransport is the same as DefaultTransport with a longer
+	// ResponseHeaderTimeout.  This is copied from DefaultTransport.  We don't
+	// do an actual copy because the Trasnport uses mutexes.  copied mutexes
+	// are ineffective.  You will panic on concurrent map access if running
+	// with other clients without instantiating a new transport for watch.
+	WatchTransport http.RoundTripper = &http.Transport{
+		DialContext: (&net.Dialer{
+			Timeout: 5 * time.Second,
+		}).DialContext,
+		DisableCompression:    true,
+		MaxIdleConns:          5,
+		MaxIdleConnsPerHost:   2,
+		IdleConnTimeout:       90 * time.Second,
+		TLSHandshakeTimeout:   5 * time.Second,
+		ResponseHeaderTimeout: 1 * time.Minute,
+		ExpectContinueTimeout: 5 * time.Second,
+	}
+)
+
+func (w *Watcher) client() *Client {
+	if client := w.Client; client != nil {
+		return client
+	}
+	c := &Client{
+		Transport: WatchTransport,
+	}
+	return c
+}
+
+// Watch is the package-level Watch defintion which is called on
+// DefaultWatcher.
+func Watch(ctx context.Context, key string, handler WatcherFunc) {
+	DefaultWatcher.Watch(ctx, key, handler)
+}
+
+// WatchPrefix is the package-level WatchPrefix defintion which is called on
+// DefaultWatcher.
+func WatchPrefix(ctx context.Context, prefix string, handler WatcherFunc) {
+	DefaultWatcher.WatchPrefix(ctx, prefix, handler)
+}
+
+// Watch executes a long poll for changes to the given key.  handler will be
+// called immediately upon registration to initialize the watch and returns
+// the initial value (as a list, this is what Consul API returns).  There is
+// a chance that Watch can miss updates due to the way the API is implemented.
+// See https://github.com/hashicorp/consul/issues/1761 for details.
+func (w *Watcher) Watch(ctx context.Context, key string, handler WatcherFunc) {
+	w.watching(ctx, key, handler, nil)
+}
+
+// WatchPrefix executes a long poll for changes to anything under the given
+// prefix.  handler will be called immediately upon registration to initialize
+// the watch and returns the initial value (as a list, this is what Consul API
+// returns).  There is a chance that WatchPrefix can miss updates due to the
+// way the API is implemented.  See https://github.com/hashicorp/consul/issues/1761
+// for details.
+func (w *Watcher) WatchPrefix(ctx context.Context, prefix string, handler WatcherFunc) {
+	q := Query{Param{
+		Name:  "recurse",
+		Value: "",
+	}}
+	w.watching(ctx, prefix, handler, q)
+}
+
+func (w *Watcher) watching(ctx context.Context, key string, handler WatcherFunc, q Query) {
+	if w.MaxAttempts <= 0 {
+		w.MaxAttempts = 10
+	}
+
+	path := "/v1/kv/" + key
+	value := "0"
+
+	attempt := 0
+	for {
+		q.Add(Param{Name: "index", Value: value})
+		resp := []KeyData{}
+		hdr, err := w.client().do(ctx, "GET", path, q, nil, &resp)
+		if err != nil {
+			if err == context.Canceled {
```

Comment:
This doesn't feel very clean to me.  Is there a better way?

Reply:
That's not gonna work because the `net/http` library wraps the errors. But you can check the context directly with `ctx.Err() != nil` for example.

Reply:
Is that the right way or is `ctx.Err() != context.Canceled` what we want?  Or, put better, will we not have to worry about context eating another error and bailing without good reason.

Reply:
This is the right way
```go
// If Done is not yet closed, Err returns nil.
// If Done is closed, Err returns a non-nil error explaining why:
// Canceled if the context was canceled
// or DeadlineExceeded if the context's deadline passed.
// After Err returns a non-nil error, successive calls to Err return the same error.
Err() error
```
You can't assume what the error will be, but if there's an error on the context it has been canceled.